cli/disputer.go

```diff
@@ -0,0 +1,359 @@
+			GasLimit:   0,
+			GasFeeCap:  abi.TokenAmount{},
+			GasPremium: abi.TokenAmount{},
+			Method:     builtin3.MethodsMiner.DisputeWindowedPoSt,
+			Params:     dpp,
+		}
+
+		rslt, err := api.StateCall(ctx, dmsg, types.EmptyTSK)
+		if err != nil {
+			return xerrors.Errorf("failed to simulate dispute: %w", err)
+		}
+
+		if rslt.MsgRct.ExitCode == 0 {
+			sm, err := api.MpoolPushMessage(ctx, dmsg, nil)
+			if err != nil {
+				return err
+			}
+
+			fmt.Println("dispute message ", sm.Cid())
+		} else {
+			fmt.Println("dispute is unsuccessful")
+		}
+
+		return nil
+	},
+}
+
+var disputerStartCmd = &cli.Command{
+	Name:      "start",
+	Usage:     "Start the window post disputer",
+	ArgsUsage: "[minerAddress]",
+	Flags: []cli.Flag{
+		&cli.StringFlag{
+			Name:  "from",
+			Usage: "optionally specify the account to send funds from",
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+		api, closer, err := GetFullNodeAPI(cctx)
+		if err != nil {
+			return err
+		}
+		defer closer()
+
+		ctx := ReqContext(cctx)
+
+		var fromAddr address.Address
+		if from := cctx.String("from"); from == "" {
+			defaddr, err := api.WalletDefaultAddress(ctx)
+			if err != nil {
+				return err
+			}
+
+			fromAddr = defaddr
+		} else {
+			addr, err := address.NewFromString(from)
+			if err != nil {
+				return err
+			}
+
+			fromAddr = addr
+		}
+
+		headChanges, err := api.ChainNotify(ctx)
+		if err != nil {
+			return err
+		}
+
+		head, ok := <-headChanges
+		if !ok {
+			return xerrors.Errorf("Notify stream was invalid")
+		}
+
+		if len(head) != 1 {
+			return xerrors.Errorf("Notify first entry should have been one item")
+		}
+
+		if head[0].Type != store.HCCurrent {
+			return xerrors.Errorf("expected current head on Notify stream (got %s)", head[0].Type)
+		}
+
+		fmt.Println("starting up window post disputer")
+
+		ticker := time.NewTicker(120 * time.Second)
+		defer ticker.Stop()
+		// TODO: Don't start unless fully synced (check syncstatus)
+		for {
+			select {
+			case notif, ok := <-headChanges:
+				if !ok {
+					return xerrors.Errorf("head change channel errored")
+				}
+				for _, val := range notif {
+					switch val.Type {
+					case store.HCApply:
+						// TODO: Confirm that we're never notified to "apply" a null tipset
+
+						if val.Val.Height() < miner3.WPoStChallengeWindow {
+							// it's too early in the chain to challenge any posts
+							continue
+						}
+
+						// TODO: If the newly applied tipset is built on top of n null tipset(s), we should also check the tipsets WPoStChallengeWindow+[1...n]  epochs ago
+
+						// Since we're waiting WPoStChallengeWindow epochs already, we don't have to worry about confidence of the message -- it's already pretty deep onchain
+						ts, err := api.ChainGetTipSetByHeight(ctx, val.Val.Height()-miner3.WPoStChallengeWindow, val.Val.Key())
+						if err != nil {
+							return xerrors.Errorf("failed to get tipset 60 epochs ago: %w", err)
+						}
+
+						if ts.IsNull() {
```

Comment:
If ChainGetTipSetByHeight lands on a nullblock it will get you the a tipset before the nullblock(s)